### PR TITLE
Move the OnCheckExemption hook out of the core.

### PR DESCRIPTION
**Pull request by SaberUK**
~~This needs further testing to check it doesn't have any hidden bugs. If anyone is interested in helping test then poke me in \#inspircd.dev.~~

### CODE REVIEW

File: src/coremods/core_channel/core_channel.cpp [hunk 1]
```diff
@@ -110,6 +134,22 @@ class CoreModChannel : public Module
 		invapi.RemoveAll(chan);
 	}
 
+	ModResult OnCheckExemption(User* user, Channel* chan, const std::string& restriction)
+	{
+		if (exemptions.count(restriction))
```
Comment (genius3000) [inline src/coremods/core_channel/core_channel.cpp]: Shouldn't this be "!exemptions.count(restriction)"?

Comment (SaberUK) [inline src/coremods/core_channel/core_channel.cpp]: Yeah, thanks.

File: src/coremods/core_channel/cmd_topic.cpp [hunk 2]
```diff
@@ -73,10 +74,15 @@ CmdResult CommandTopic::HandleLocal(const std::vector<std::string>& parameters,
...
 			return CMD_FAILURE;
 		}
-		if (c->IsModeSet(topiclockmode) && !ServerInstance->OnCheckExemption(user, c, "topiclock").check(c->GetPrefixValue(user) >= HALFOP_VALUE))
+		if (c->IsModeSet(topiclockmode))
 		{
-			user->WriteNumeric(ERR_CHANOPRIVSNEEDED, c->name, "You do not have access to change the topic on this channel");
-			return CMD_FAILURE;
+			ModResult MOD_RESULT;
+			FIRST_MOD_RESULT_CUSTOM(*this, CheckExemption::EventListener, OnCheckExemption, MOD_RESULT, (user, c, "topiclock"));
+			if (MOD_RESULT.check(c->GetPrefixValue(user) >= HALFOP_VALUE))
```
Comment (genius3000) [inline src/coremods/core_channel/cmd_topic.cpp]: This if statement needs to be inverted as well.

**Comment by genius3000** [comment]
@SaberUK Other than those two inverted if statements, I ran through each mode and exemption; everything seems to work as expected. Can't guarantee I didn't miss anything, but each exemption worked at the level (and above) that I told it to, so looks good.

File: include/modules/exemption.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	EventListener(Module* mod)
+		: ModuleEventListener(mod, "event/exemption")
+	{
+	}
+
+	/** Called when checking if a user has exemption from something.
+	 * @param user The user to check exemption for.
+	 * @param chan The channel to check exemption on.
+	 * @param restriction The restriction to check for.
+	 * @return ...
```
Comment (attilamolnar) [inline include/modules/exemption.h]: Documentation missing from here

**Comment by attilamolnar** [comment]
I'd prefer if you used composition instead of inheritance for the event providers, that way the affected modules can provide multiple events if needed in the future and remain uniform

**Comment by SaberUK** [comment]
@attilamolnar Done.

**Comment by SaberUK** [comment]
@genius3000 Thanks!

**Pull request merged**